### PR TITLE
Add class_skip_list config to ir_type_checker

### PR DESCRIPTION
**Pull request by justhecuke**
We have a use-case where our build has some poorly behaved bytecode manipulators that change things in a way that IR Type Checker does not like. There is no easy to way to fix them but they do not cause runtime errors so the easiest thing for us to do is to use a skip list for IR Type Checker to allow us to continue even if it finds things it doesn't like.

Example:
redex.config
```
{
  "ir_type_checker": {
    "class_skip_list": [
      "Lcom/some/class;",
      "Lcom/some/other/class;"
    ]
  }
}
```

### CODE REVIEW

**Comment by yuxuanchen1997** [comment]
Can you provide an example of bytecode where it's rejected by the `IRTypeChecker` and accepted by Android Runtime? Code needs to be able to pass `IRTypeChecker` in order to make sure the optimization passes are actually correct.

File: libredex/IRTypeChecker.h [hunk 1]
```diff
@@ -32,7 +32,10 @@ class IRTypeChecker final {
   // definition must be located after the definition of TypeInference.
   ~IRTypeChecker();
 
-  explicit IRTypeChecker(DexMethod* dex_method, bool validate_access = false);
+  explicit IRTypeChecker(DexMethod* dex_method,
+                         bool validate_access = false,
+                         std::unordered_set<std::string> class_skip_list = {
+                             ""});
```
Comment (yuxuanchen1997) [inline libredex/IRTypeChecker.h:38]: Why the empty string?

File: libredex/PassManager.cpp [hunk 2]
```diff
@@ -201,13 +207,14 @@ struct CheckerConfig {
                                                    bool verify_moves,
                                                    bool check_no_overwrite_this,
                                                    bool validate_access,
+                                                   const std::unordered_set<std::string> class_skip_list,
```
Comment (yuxuanchen1997) [inline libredex/PassManager.cpp:210]:
```suggestion
                                                   const std::unordered_set<std::string>& class_skip_list,
```

File: libredex/IRTypeChecker.cpp [hunk 3]
```diff
@@ -656,14 +656,15 @@ void validate_access(const DexMethod* accessor, const DexMember* accessee) {
 
 IRTypeChecker::~IRTypeChecker() {}
 
-IRTypeChecker::IRTypeChecker(DexMethod* dex_method, bool validate_access)
+IRTypeChecker::IRTypeChecker(DexMethod* dex_method, bool validate_access, std::unordered_set<std::string> class_skip_list)
```
Comment (yuxuanchen1997) [inline libredex/IRTypeChecker.cpp:659]:
```suggestion
IRTypeChecker::IRTypeChecker(DexMethod* dex_method, bool validate_access, const std::unordered_set<std::string>& class_skip_list)
```

File: libredex/IRTypeChecker.h [hunk 4]
```diff
@@ -32,7 +32,10 @@ class IRTypeChecker final {
   // definition must be located after the definition of TypeInference.
   ~IRTypeChecker();
 
-  explicit IRTypeChecker(DexMethod* dex_method, bool validate_access = false);
+  explicit IRTypeChecker(DexMethod* dex_method,
+                         bool validate_access = false,
+                         std::unordered_set<std::string> class_skip_list = {
```
Comment (yuxuanchen1997) [inline libredex/IRTypeChecker.h:37]:
```suggestion
                         const std::unordered_set<std::string>& class_skip_list = {
```

**Review by yuxuanchen1997 — CHANGES_REQUESTED** [review]

File: libredex/IRTypeChecker.h (hunk 1 above)
Comment (justhecuke) [reply libredex/IRTypeChecker.h]: Artifact of trying to get this to work. No real reason to keep.

**Comment by justhecuke** [comment]
> Can you provide an example of bytecode where it's rejected by the `IRTypeChecker` and accepted by Android Runtime? Code needs to be able to pass `IRTypeChecker` in order to make sure the optimization passes are actually correct.

I think, from the bytecode, that we're injecting a wrapper class into RecyclerView's package to access package-private fields but IRTypeChecker doesn't like that. Honestly, seems like some sort of bug since the calling class is in the same package.

[PM:1] Running IRTypeChecker...
Inconsistency found in Dex code for Landroidx/appcompat/widget/t;.a:(Landroidx/recyclerview/widget/RecyclerView;I)I
 Type error in method Landroidx/appcompat/widget/WrappedRecyclerHelperKt;.getRecyclerMaxScrap:(Landroidx/recyclerview/widget/RecyclerView;I)I at instruction 'IGET_OBJECT v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;.mScrap:Landroid/util/SparseArray;' @ 0x7f8e9e062ca0 for 
illegal access to package-private Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;.mScrap:Landroid/util/SparseArray;
 from Landroidx/appcompat/widget/WrappedRecyclerHelperKt;.getRecyclerMaxScrap:(Landroidx/recyclerview/widget/RecyclerView;I)I
Code:
[0x7f8e9e05e070] OPCODE: IOPCODE_LOAD_PARAM_OBJECT v1
[0x7f8e9e061f70] OPCODE: IOPCODE_LOAD_PARAM v2
[0x7f8e9e061f90] OPCODE: CONST_STRING "recyclerView"
[0x7f8e4e04fc40] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO_OBJECT v0
[0x7f8e9e061890] OPCODE: INVOKE_STATIC v1, v0, Lg/f/b/m;.b:(Ljava/lang/Object;Ljava/lang/String;)V
[0x7f8e9e035ae0] POSITION: Landroidx/appcompat/widget/WrappedRecyclerHelperKt;.getRecyclerMaxScrap:(Landroidx/recyclerview/widget/RecyclerView;I)I(WrappedRecyclerHelperKt.java:10)
[0x7f8e9e0618b0] OPCODE: INVOKE_VIRTUAL v1, Landroidx/recyclerview/widget/RecyclerView;.getRecycledViewPool:()Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;
[0x7f8e9e062c80] OPCODE: MOVE_RESULT_OBJECT v0
[0x7f8e9e062ca0] OPCODE: IGET_OBJECT v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;.mScrap:Landroid/util/SparseArray;
[0x7f8e4e050130] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO_OBJECT v0
[0x7f8e4e04f8a0] OPCODE: INVOKE_VIRTUAL v0, v2, Landroid/util/SparseArray;.get:(I)Ljava/lang/Object;
[0x7f8e4e04f8c0] OPCODE: MOVE_RESULT_OBJECT v0
[0x7f8e4e04faa0] OPCODE: CHECK_CAST v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool$ScrapData;
[0x7f8e6b407320] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO_OBJECT v0
[0x7f8e4e04fac0] OPCODE: IF_EQZ v0
[0x7f8e4e051710] OPCODE: IGET v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool$ScrapData;.mMaxScrap:I
[0x7f8e6b406030] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO v0
[0x7f8e4e051730] OPCODE: RETURN v0
[0x7f8de877e990] TARGET: SIMPLE 0x7f8e4e04fac0
[0x7f8e4e04f720] OPCODE: CONST v0, -1
[0x7f8e4e04f740] OPCODE: RETURN v0

**Comment by justhecuke** [comment]
> > Can you provide an example of bytecode where it's rejected by the `IRTypeChecker` and accepted by Android Runtime? Code needs to be able to pass `IRTypeChecker` in order to make sure the optimization passes are actually correct.
> 
> I think, from the bytecode, that we're injecting a wrapper class into RecyclerView's package to access package-private fields but IRTypeChecker doesn't like that. Honestly, seems like some sort of bug since the calling class is in the same package.
> 
> [PM:1] Running IRTypeChecker...
> Inconsistency found in Dex code for Landroidx/appcompat/widget/t;.a:(Landroidx/recyclerview/widget/RecyclerView;I)I
> Type error in method Landroidx/appcompat/widget/WrappedRecyclerHelperKt;.getRecyclerMaxScrap:(Landroidx/recyclerview/widget/RecyclerView;I)I at instruction 'IGET_OBJECT v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;.mScrap:Landroid/util/SparseArray;' @ 0x7f8e9e062ca0 for
> illegal access to package-private Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;.mScrap:Landroid/util/SparseArray;
> from Landroidx/appcompat/widget/WrappedRecyclerHelperKt;.getRecyclerMaxScrap:(Landroidx/recyclerview/widget/RecyclerView;I)I
> Code:
> [0x7f8e9e05e070] OPCODE: IOPCODE_LOAD_PARAM_OBJECT v1
> [0x7f8e9e061f70] OPCODE: IOPCODE_LOAD_PARAM v2
> [0x7f8e9e061f90] OPCODE: CONST_STRING "recyclerView"
> [0x7f8e4e04fc40] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO_OBJECT v0
> [0x7f8e9e061890] OPCODE: INVOKE_STATIC v1, v0, Lg/f/b/m;.b:(Ljava/lang/Object;Ljava/lang/String;)V
> [0x7f8e9e035ae0] POSITION: Landroidx/appcompat/widget/WrappedRecyclerHelperKt;.getRecyclerMaxScrap:(Landroidx/recyclerview/widget/RecyclerView;I)I(WrappedRecyclerHelperKt.java:10)
> [0x7f8e9e0618b0] OPCODE: INVOKE_VIRTUAL v1, Landroidx/recyclerview/widget/RecyclerView;.getRecycledViewPool:()Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;
> [0x7f8e9e062c80] OPCODE: MOVE_RESULT_OBJECT v0
> [0x7f8e9e062ca0] OPCODE: IGET_OBJECT v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool;.mScrap:Landroid/util/SparseArray;
> [0x7f8e4e050130] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO_OBJECT v0
> [0x7f8e4e04f8a0] OPCODE: INVOKE_VIRTUAL v0, v2, Landroid/util/SparseArray;.get:(I)Ljava/lang/Object;
> [0x7f8e4e04f8c0] OPCODE: MOVE_RESULT_OBJECT v0
> [0x7f8e4e04faa0] OPCODE: CHECK_CAST v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool$ScrapData;
> [0x7f8e6b407320] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO_OBJECT v0
> [0x7f8e4e04fac0] OPCODE: IF_EQZ v0
> [0x7f8e4e051710] OPCODE: IGET v0, Landroidx/recyclerview/widget/RecyclerView$RecycledViewPool$ScrapData;.mMaxScrap:I
> [0x7f8e6b406030] OPCODE: IOPCODE_MOVE_RESULT_PSEUDO v0
> [0x7f8e4e051730] OPCODE: RETURN v0
> [0x7f8de877e990] TARGET: SIMPLE 0x7f8e4e04fac0
> [0x7f8e4e04f720] OPCODE: CONST v0, -1
> [0x7f8e4e04f740] OPCODE: RETURN v0

Oops, missed that the packages are, indeed, different. My eyes saw the similar class name and skipped over the s/appcompat/widget/ part.

**Comment by justhecuke** [comment]
@yuxuanchen1997 Let me know if you need anything else. Our setup is very strange so I wouldn't be surprised if redex reacts in new and surprising ways because of that.

**Comment by yuxuanchen1997** [comment]
@justhecuke My main concern here is that this patch doesn't seem to solve any particular problems. Lots of Redex code are dependent on IRTypeChecker passing. The example you provided is exactly why we need IRTypeChecker in the first place. Feel free to reopen if my judgement is wrong.

**Comment by justhecuke** [comment]
The use-case is to allow the user to get their APK to work with Redex. Without this, we're going to have to maintain our own fork in order to use redex on our APK which, while survivable, will undoubtedly be annoying.

**Comment by yuxuanchen1997** [comment]
This doesn't fix the problem. I would suggest that you fix the issue from where it originated. In your example, the access could lead to `IllegalAccessError`s at runtime. We should see the `IRTypeChecker` enforcement as a contract between passes so that any pass can safely assume the program meets certain specifications. That's also why we provide an option to run `IRTypeChecker` after every pass (and you should expect no errors between passes). If we allow holes to be poked in `IRTypeChecker`, it will make writing passes a lot more difficult and profound bugs to occur more often. It's similar to how we don't just remove assertions in our program when they fail. 

If for some reasons your issue can't be fixed from the origin, you can try to use `MakePublicPass` and see if that makes the type error disappear.

**Comment by justhecuke** [comment]
As I mentioned before, fixing the previous bytecode manipulations is difficult and they don't actually result in runtime errors. I'm not sure how they managed that, but they did.

While I appreciate where you're coming from, it's impractical on the time scale that led us to use redex in the first place. Not to mention that it's a separate team in a separate org that actually owns those things which adds extra layers of complication. It's on the radar to do, but it'll take a while to get to.

I'll give your suggestion to use `MakePublicPass` a try.

**Comment by amnn** [comment]
Hi @justhecuke,

Am I right in thinking that the code that's accessing the package private variable is a Kotlin extension function, defined something like this?

```kotlin
package androidx.appcompat.widget

import androidx.recyclerview.widget.RecyclerView

fun RecyclerView.getRecyclerMaxScrap(x: Int): Int {
   /* ... */
}
```

If so, would it be feasible for you to move it into the same package as `RecyclerView`?  I have not tested it, but suspect that would address the issue without resorting to compiler intervention.

**Comment by justhecuke** [comment]
> Hi @justhecuke,
> 
> Am I right in thinking that the code that's accessing the package private variable is a Kotlin extension function, defined something like this?
> 
> ```kotlin
> package androidx.appcompat.widget
> 
> import androidx.recyclerview.widget.RecyclerView
> 
> fun RecyclerView.getRecyclerMaxScrap(x: Int): Int {
>    /* ... */
> }
> ```
> 
> If so, would it be feasible for you to move it into the same package as `RecyclerView`? I have not tested it, but suspect that would address the issue without resorting to compiler intervention.

It appears to just be a regular kotlin function that simply uses the package-private field directly and isn't an extension function. But, it does seem to go through a bytecode transform of some sort because the pre-redex class file does not match the source code I'm looking at.

**Comment by amnn** [comment]
In any case, if this function were defined within the same package as `RecyclerView`, (which I believe is the convention for such Kotlin extension libraries), then that would address your issue without the help of a compiler making things public.

**Pull request closed without merging**